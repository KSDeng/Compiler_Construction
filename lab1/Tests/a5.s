.data
_prompt: .asciiz "Enter an integer:"
_ret: .asciiz "\n"
.globl main
.text

read:
  li $v0, 4
  la $a0, _prompt
  syscall
  li $v0, 5
  syscall
  jr $ra

write:
  li $v0, 1
  syscall
  li $v0, 4
  la $a0, _ret
  syscall
  move $v0, $0
  jr $ra


compare:
  move $t0, $a0
  move $t1, $a1
  bgt $t0, $t1, label0
  j label1
label0:
  move $v0, $t0
  jr $ra
label1:
  blt $t0, $t1, label2
  j label3
label2:
  move $v0, $t1
  jr $ra
label3:
  move $v0, $0
  jr $ra

add2:
  move $t0, $a0
  move $t1, $a1
  add $t2, $t0, $t1
  move $v0, $t2
  jr $ra

main:
  li $t0, 0
label4:
  li $t1, 10
  blt $t0, $t1, label5
  j label6
label5:
  addi $sp, $sp, -4
  sw $ra, 0($sp)
  jal read
  lw $ra, 0($sp)
  addi $sp, $sp, 4
  move $t2, $v0
  move $t3, $t0
  li $t5, 4
  mul $t4, $t3, $t5
  lw $t6, 0($t7)
  add $s0, $t6, $t4
  sw $s0, 0($t2)
  addi $s1, $t0, 1
  move $t0, $s1
  j label4
label6:
  li $t0, 0
  li $s2, 0
label7:
  li $t1, 10
  blt $t0, $t1, label8
  j label9
label8:
  move $s3, $t0
  li $t5, 4
  mul $s4, $s3, $t5
  lw $s5, 0($t7)
  add $s6, $s5, $s4
  lw $s7, 0($s6)
  addi $t8, $t0, 1
  li $t5, 4
  mul $t9, $t8, $t5
  lw $t3, 0($t7)
  add $t6, $t3, $t9
  lw $t4, 0($t6)
  addi $sp, $sp, -12
  sw $ra, 0($sp)
  sw $a0, 4($sp)
  sw $a1, 8($sp)
  move $a0, $s7
  move $a1, $t4
  addi $sp, $sp, -4
  sw $t0, 0($sp)
  addi $sp, $sp, -4
  sw $t1, 0($sp)
  addi $sp, $sp, -4
  sw $t2, 0($sp)
  addi $sp, $sp, -4
  sw $t3, 0($sp)
  addi $sp, $sp, -4
  sw $t4, 0($sp)
  addi $sp, $sp, -4
  sw $t5, 0($sp)
  addi $sp, $sp, -4
  sw $t6, 0($sp)
  addi $sp, $sp, -4
  sw $t7, 0($sp)
  addi $sp, $sp, -4
  sw $s0, 0($sp)
  addi $sp, $sp, -4
  sw $s1, 0($sp)
  addi $sp, $sp, -4
  sw $s2, 0($sp)
  addi $sp, $sp, -4
  sw $s3, 0($sp)
  addi $sp, $sp, -4
  sw $s4, 0($sp)
  addi $sp, $sp, -4
  sw $s5, 0($sp)
  addi $sp, $sp, -4
  sw $s6, 0($sp)
  addi $sp, $sp, -4
  sw $s7, 0($sp)
  addi $sp, $sp, -4
  sw $t8, 0($sp)
  addi $sp, $sp, -4
  sw $t9, 0($sp)
  jal compare
  lw $t9, 0($sp)
  addi $sp, $sp, 4
  lw $t8, 0($sp)
  addi $sp, $sp, 4
  lw $s7, 0($sp)
  addi $sp, $sp, 4
  lw $s6, 0($sp)
  addi $sp, $sp, 4
  lw $s5, 0($sp)
  addi $sp, $sp, 4
  lw $s4, 0($sp)
  addi $sp, $sp, 4
  lw $s3, 0($sp)
  addi $sp, $sp, 4
  lw $s2, 0($sp)
  addi $sp, $sp, 4
  lw $s1, 0($sp)
  addi $sp, $sp, 4
  lw $s0, 0($sp)
  addi $sp, $sp, 4
  lw $t7, 0($sp)
  addi $sp, $sp, 4
  lw $t6, 0($sp)
  addi $sp, $sp, 4
  lw $t5, 0($sp)
  addi $sp, $sp, 4
  lw $t4, 0($sp)
  addi $sp, $sp, 4
  lw $t3, 0($sp)
  addi $sp, $sp, 4
  lw $t2, 0($sp)
  addi $sp, $sp, 4
  lw $t1, 0($sp)
  addi $sp, $sp, 4
  lw $t0, 0($sp)
  addi $sp, $sp, 4
  lw $a1, 8($sp)
  lw $a0, 4($sp)
  lw $ra, 0($sp)
  addi $sp, $sp, 12
  move $t0, $v0
  move $t0, $s2
  li $t5, 4
  mul $t0, $t1, $t5
  lw $t2, 0($t3)
  add $t4, $t2, $t0
  sw $t4, 0($t6)
  addi $s0, $t7, 2
  move $t7, $s0
  addi $s1, $s2, 1
  move $s2, $s1
  j label7
label9:
  li $t7, 0
label10:
  li $s3, 5
  blt $t7, $s3, label11
  j label12
label11:
  move $s4, $t7
  li $t5, 4
  mul $s5, $s4, $t5
  lw $s6, 0($t3)
  add $s7, $s6, $s5
  lw $t8, 0($s7)
  move $t9, $t7
  li $t5, 4
  mul $t1, $t9, $t5
  lw $t2, 0($t0)
  add $t4, $t2, $t1
  lw $t6, 0($t4)
  addi $sp, $sp, -12
  sw $ra, 0($sp)
  sw $a0, 4($sp)
  sw $a1, 8($sp)
  move $a0, $t8
  move $a1, $t6
  addi $sp, $sp, -4
  sw $t0, 0($sp)
  addi $sp, $sp, -4
  sw $t1, 0($sp)
  addi $sp, $sp, -4
  sw $t2, 0($sp)
  addi $sp, $sp, -4
  sw $t3, 0($sp)
  addi $sp, $sp, -4
  sw $t4, 0($sp)
  addi $sp, $sp, -4
  sw $t5, 0($sp)
  addi $sp, $sp, -4
  sw $t6, 0($sp)
  addi $sp, $sp, -4
  sw $t7, 0($sp)
  addi $sp, $sp, -4
  sw $s0, 0($sp)
  addi $sp, $sp, -4
  sw $s1, 0($sp)
  addi $sp, $sp, -4
  sw $s2, 0($sp)
  addi $sp, $sp, -4
  sw $s3, 0($sp)
  addi $sp, $sp, -4
  sw $s4, 0($sp)
  addi $sp, $sp, -4
  sw $s5, 0($sp)
  addi $sp, $sp, -4
  sw $s6, 0($sp)
  addi $sp, $sp, -4
  sw $s7, 0($sp)
  addi $sp, $sp, -4
  sw $t8, 0($sp)
  addi $sp, $sp, -4
  sw $t9, 0($sp)
  jal add2
  lw $t9, 0($sp)
  addi $sp, $sp, 4
  lw $t8, 0($sp)
  addi $sp, $sp, 4
  lw $s7, 0($sp)
  addi $sp, $sp, 4
  lw $s6, 0($sp)
  addi $sp, $sp, 4
  lw $s5, 0($sp)
  addi $sp, $sp, 4
  lw $s4, 0($sp)
  addi $sp, $sp, 4
  lw $s3, 0($sp)
  addi $sp, $sp, 4
  lw $s2, 0($sp)
  addi $sp, $sp, 4
  lw $s1, 0($sp)
  addi $sp, $sp, 4
  lw $s0, 0($sp)
  addi $sp, $sp, 4
  lw $t7, 0($sp)
  addi $sp, $sp, 4
  lw $t6, 0($sp)
  addi $sp, $sp, 4
  lw $t5, 0($sp)
  addi $sp, $sp, 4
  lw $t4, 0($sp)
  addi $sp, $sp, 4
  lw $t3, 0($sp)
  addi $sp, $sp, 4
  lw $t2, 0($sp)
  addi $sp, $sp, 4
  lw $t1, 0($sp)
  addi $sp, $sp, 4
  lw $t0, 0($sp)
  addi $sp, $sp, 4
  lw $a1, 8($sp)
  lw $a0, 4($sp)
  lw $ra, 0($sp)
  addi $sp, $sp, 12
  move $t0, $v0
  move $t0, $t7
  li $t5, 4
  mul $t0, $t1, $t5
  lw $t2, 0($t3)
  add $t4, $t2, $t0
  sw $t4, 0($t6)
  move $s0, $t7
  li $t5, 4
  mul $s1, $s0, $t5
  lw $s2, 0($t3)
  add $s3, $s2, $s1
  lw $s4, 0($s3)
  move $a0, $s4
  addi $sp, $sp, -4
  sw $ra, 0($sp)
  jal write
  lw $ra, 0($sp)
  addi $sp, $sp, 4
  addi $s5, $t7, 1
  move $t7, $s5
  j label10
label12:
  move $v0, $0
  jr $ra
