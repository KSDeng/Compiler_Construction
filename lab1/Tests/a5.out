Program (1)
  ExtDefList (1)
    ExtDef (1)
      Specifier (1)
        TYPE: int
      FunDec (1)
        ID: compare
        LP
        VarList (1)
          ParamDec (1)
            Specifier (1)
              TYPE: int
            VarDec (1)
              ID: a
          COMMA
          VarList (1)
            ParamDec (1)
              Specifier (1)
                TYPE: int
              VarDec (1)
                ID: b
        RP
      CompSt (1)
        LC
        StmtList (2)
          Stmt (2)
            IF
            LP
            Exp (2)
              Exp (2)
                ID: a
              RELOP
              Exp (2)
                ID: b
            RP
            Stmt (2)
              RETURN
              Exp (2)
                ID: a
              SEMI
          StmtList (3)
            Stmt (3)
              IF
              LP
              Exp (3)
                Exp (3)
                  ID: a
                RELOP
                Exp (3)
                  ID: b
              RP
              Stmt (3)
                RETURN
                Exp (3)
                  ID: b
                SEMI
            StmtList (4)
              Stmt (4)
                RETURN
                Exp (4)
                  INT: 0
                SEMI
        RC
    ExtDefList (7)
      ExtDef (7)
        Specifier (7)
          TYPE: int
        FunDec (7)
          ID: add2
          LP
          VarList (7)
            ParamDec (7)
              Specifier (7)
                TYPE: int
              VarDec (7)
                ID: aa
            COMMA
            VarList (7)
              ParamDec (7)
                Specifier (7)
                  TYPE: int
                VarDec (7)
                  ID: bb
          RP
        CompSt (7)
          LC
          StmtList (8)
            Stmt (8)
              RETURN
              Exp (8)
                Exp (8)
                  ID: aa
                PLUS
                Exp (8)
                  ID: bb
              SEMI
          RC
      ExtDefList (11)
        ExtDef (11)
          Specifier (11)
            TYPE: int
          FunDec (11)
            ID: main
            LP
            RP
          CompSt (11)
            LC
            DefList (12)
              Def (12)
                Specifier (12)
                  TYPE: int
                DecList (12)
                  Dec (12)
                    VarDec (12)
                      VarDec (12)
                        ID: i
                      LB
                      INT: 10
                      RB
                  COMMA
                  DecList (12)
                    Dec (12)
                      VarDec (12)
                        VarDec (12)
                          ID: r
                        LB
                        INT: 5
                        RB
                    COMMA
                    DecList (12)
                      Dec (12)
                        VarDec (12)
                          ID: n
                      COMMA
                      DecList (12)
                        Dec (12)
                          VarDec (12)
                            ID: m
                SEMI
            StmtList (13)
              Stmt (13)
                Exp (13)
                  Exp (13)
                    ID: n
                  ASSIGNOP
                  Exp (13)
                    INT: 0
                SEMI
              StmtList (14)
                Stmt (14)
                  WHILE
                  LP
                  Exp (14)
                    Exp (14)
                      ID: n
                    RELOP
                    Exp (14)
                      INT: 10
                  RP
                  Stmt (14)
                    CompSt (14)
                      LC
                      StmtList (15)
                        Stmt (15)
                          Exp (15)
                            Exp (15)
                              Exp (15)
                                ID: i
                              LB
                              Exp (15)
                                ID: n
                              RB
                            ASSIGNOP
                            Exp (15)
                              ID: read
                              LP
                              RP
                          SEMI
                        StmtList (16)
                          Stmt (16)
                            Exp (16)
                              Exp (16)
                                ID: n
                              ASSIGNOP
                              Exp (16)
                                Exp (16)
                                  ID: n
                                PLUS
                                Exp (16)
                                  INT: 1
                            SEMI
                      RC
                StmtList (18)
                  Stmt (18)
                    Exp (18)
                      Exp (18)
                        ID: n
                      ASSIGNOP
                      Exp (18)
                        INT: 0
                    SEMI
                  StmtList (19)
                    Stmt (19)
                      Exp (19)
                        Exp (19)
                          ID: m
                        ASSIGNOP
                        Exp (19)
                          INT: 0
                      SEMI
                    StmtList (20)
                      Stmt (20)
                        WHILE
                        LP
                        Exp (20)
                          Exp (20)
                            ID: n
                          RELOP
                          Exp (20)
                            INT: 10
                        RP
                        Stmt (20)
                          CompSt (20)
                            LC
                            StmtList (21)
                              Stmt (21)
                                Exp (21)
                                  Exp (21)
                                    Exp (21)
                                      ID: r
                                    LB
                                    Exp (21)
                                      ID: m
                                    RB
                                  ASSIGNOP
                                  Exp (21)
                                    ID: compare
                                    LP
                                    Args (21)
                                      Exp (21)
                                        Exp (21)
                                          ID: i
                                        LB
                                        Exp (21)
                                          ID: n
                                        RB
                                      COMMA
                                      Args (21)
                                        Exp (21)
                                          Exp (21)
                                            ID: i
                                          LB
                                          Exp (21)
                                            Exp (21)
                                              ID: n
                                            PLUS
                                            Exp (21)
                                              INT: 1
                                          RB
                                    RP
                                SEMI
                              StmtList (22)
                                Stmt (22)
                                  Exp (22)
                                    Exp (22)
                                      ID: n
                                    ASSIGNOP
                                    Exp (22)
                                      Exp (22)
                                        ID: n
                                      PLUS
                                      Exp (22)
                                        INT: 2
                                  SEMI
                                StmtList (23)
                                  Stmt (23)
                                    Exp (23)
                                      Exp (23)
                                        ID: m
                                      ASSIGNOP
                                      Exp (23)
                                        Exp (23)
                                          ID: m
                                        PLUS
                                        Exp (23)
                                          INT: 1
                                    SEMI
                            RC
                      StmtList (25)
                        Stmt (25)
                          Exp (25)
                            Exp (25)
                              ID: n
                            ASSIGNOP
                            Exp (25)
                              INT: 0
                          SEMI
                        StmtList (26)
                          Stmt (26)
                            WHILE
                            LP
                            Exp (26)
                              Exp (26)
                                ID: n
                              RELOP
                              Exp (26)
                                INT: 5
                            RP
                            Stmt (26)
                              CompSt (26)
                                LC
                                StmtList (27)
                                  Stmt (27)
                                    Exp (27)
                                      Exp (27)
                                        Exp (27)
                                          ID: r
                                        LB
                                        Exp (27)
                                          ID: n
                                        RB
                                      ASSIGNOP
                                      Exp (27)
                                        ID: add2
                                        LP
                                        Args (27)
                                          Exp (27)
                                            Exp (27)
                                              ID: r
                                            LB
                                            Exp (27)
                                              ID: n
                                            RB
                                          COMMA
                                          Args (27)
                                            Exp (27)
                                              Exp (27)
                                                ID: i
                                              LB
                                              Exp (27)
                                                ID: n
                                              RB
                                        RP
                                    SEMI
                                  StmtList (28)
                                    Stmt (28)
                                      Exp (28)
                                        ID: write
                                        LP
                                        Args (28)
                                          Exp (28)
                                            Exp (28)
                                              ID: r
                                            LB
                                            Exp (28)
                                              ID: n
                                            RB
                                        RP
                                      SEMI
                                    StmtList (29)
                                      Stmt (29)
                                        Exp (29)
                                          Exp (29)
                                            ID: n
                                          ASSIGNOP
                                          Exp (29)
                                            Exp (29)
                                              ID: n
                                            PLUS
                                            Exp (29)
                                              INT: 1
                                        SEMI
                                RC
                          StmtList (31)
                            Stmt (31)
                              RETURN
                              Exp (31)
                                INT: 0
                              SEMI
            RC
